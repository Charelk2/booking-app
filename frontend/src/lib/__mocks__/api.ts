const handler: ProxyHandler<any> = {
  get: (target, prop) => {
    if (!(prop in target)) {
      target[prop] = jest.fn();
    }
    return target[prop];
  },
};

const api: any = new Proxy(
  { getServiceCategories: jest.fn().mockResolvedValue({ data: [] }) },
  handler,
);

module.exports = api;
