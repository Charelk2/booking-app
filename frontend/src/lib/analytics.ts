export type AnalyticsPayload = Record<string, unknown>;

export const trackEvent = (event: string, payload: AnalyticsPayload = {}): void => {
  if (typeof window !== 'undefined') {
    const anyWindow = window as any;
    if (anyWindow.analytics && typeof anyWindow.analytics.track === 'function') {
      anyWindow.analytics.track(event, payload);
      return;
    }
  }
  // Fallback to console so events are still visible during development
  // eslint-disable-next-line no-console
  console.info('analytics event', event, payload);
};
