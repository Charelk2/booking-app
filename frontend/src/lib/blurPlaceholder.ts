export const BLUR_PLACEHOLDER =
  'data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMSIgaGVpZ2h0PSIxIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxyZWN0IHdpZHRoPSIxIiBoZWlnaHQ9IjEiIGZpbGw9IiNlZWVlZWUiIC8+PC9zdmc+';
