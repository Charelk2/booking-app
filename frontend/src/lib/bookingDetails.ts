export interface ParsedBookingDetails {
  eventType?: string;
  description?: string;
  date?: string;
  location?: string;
  guests?: string;
  venueType?: string;
  soundNeeded?: string;
  notes?: string;
}

import { BOOKING_DETAILS_PREFIX } from './constants';

export function parseBookingDetailsFromMessage(content: string): ParsedBookingDetails {
  const details: ParsedBookingDetails = {};
  const lines = content.replace(BOOKING_DETAILS_PREFIX, '').trim().split('\n');
  lines.forEach((line) => {
    const parts = line.split(':');
    if (parts.length >= 2) {
      const key = parts[0].trim();
      const value = parts.slice(1).join(':').trim();
      switch (key) {
        case 'Event Type':
          details.eventType = value;
          break;
        case 'Description':
          details.description = value;
          break;
        case 'Date':
          details.date = value;
          break;
        case 'Location':
          details.location = value;
          break;
        case 'Guests':
          details.guests = value;
          break;
        case 'Venue':
          details.venueType = value;
          break;
        case 'Sound':
          details.soundNeeded = value;
          break;
        case 'Notes':
          details.notes = value;
          break;
        default:
          break;
      }
    }
  });
  return details;
}
