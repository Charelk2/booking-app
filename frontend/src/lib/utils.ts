import api from './api';

export const getFullImageUrl = (relativePath: string | undefined | null): string | null => {
  if (!relativePath) return null;
  if (relativePath.startsWith('http://') || relativePath.startsWith('https://')) {
    return relativePath;
  }
  const cleanPath = relativePath.startsWith('/static/') ? relativePath : `/static/${relativePath.replace(/^\/+/, '')}`;
  return `${api.defaults.baseURL}${cleanPath}`;
};

export const extractErrorMessage = (detail: unknown): string => {
  if (!detail) return 'An unexpected error occurred.';
  if (typeof detail === 'string') return detail;
  if (Array.isArray(detail)) {
    return detail
      .map((d) => {
        if (typeof d === 'object' && d && 'msg' in d) {
          const record = d as Record<string, unknown>;
          return record.msg as string;
        }
        return JSON.stringify(d);
      })
      .join(', ');
  }
  if (typeof detail === 'object') {
    const record = detail as Record<string, unknown>;
    if (typeof record.msg === 'string') return record.msg;
    return JSON.stringify(detail);
  }
  return String(detail);
};
