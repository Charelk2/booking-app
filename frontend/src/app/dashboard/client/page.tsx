"use client";

import React, { useEffect, useState, useMemo } from "react";
import { useRouter, usePathname, useSearchParams } from "next/navigation";
import MainLayout from "@/components/layout/MainLayout";
import { useAuth } from "@/contexts/AuthContext";
import { Booking, BookingRequest } from "@/types";
import { getMyClientBookings, getMyBookingRequests } from "@/lib/api";
import SectionList from "@/components/dashboard/SectionList";
import BookingRequestCard from "@/components/dashboard/BookingRequestCard";
import { Spinner } from "@/components/ui";
import DashboardTabs from "@/components/dashboard/DashboardTabs";
import { format } from "date-fns";
import { formatCurrency, formatStatus } from "@/lib/utils";
import Link from "next/link";

export default function ClientDashboardPage() {
  const { user, loading: authLoading } = useAuth();
  const router = useRouter();
  const pathname = usePathname();
  const params = useSearchParams();

  const [bookings, setBookings] = useState<Booking[]>([]);
  const [bookingRequests, setBookingRequests] = useState<BookingRequest[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState("");

  const initialTab = params.get("tab") === "bookings" ? "bookings" : "requests";
  const [activeTab, setActiveTab] = useState<"requests" | "bookings">(initialTab);

  useEffect(() => {
    const params = new URLSearchParams(window.location.search);
    params.set("tab", activeTab);
    router.replace(`${pathname}?${params.toString()}`);
  }, [activeTab, router, pathname]);

  useEffect(() => {
    if (authLoading) return;
    if (!user) {
      router.push(`/login?next=${encodeURIComponent(pathname)}`);
      return;
    }
    if (user.user_type !== "client") {
      router.push("/dashboard/artist");
      return;
    }

    const fetchData = async () => {
      try {
        const [bookingsData, requestsData] = await Promise.all([
          getMyClientBookings(),
          getMyBookingRequests(),
        ]);
        setBookings(bookingsData.data);
        setBookingRequests(requestsData.data);
      } catch (err) {
        console.error("Client dashboard fetch error:", err);
        setError("Failed to load dashboard data. Please try again.");
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [user, authLoading, router, pathname]);

  const upcomingBookings = useMemo(() => {
    const now = new Date().getTime();
    return bookings
      .filter((b) => new Date(b.start_time).getTime() >= now)
      .sort((a, b) => new Date(a.start_time).getTime() - new Date(b.start_time).getTime())
      .slice(0, 5);
  }, [bookings]);

  if (loading) {
    return (
      <MainLayout>
        <div className="p-8 flex justify-center"><Spinner /></div>
      </MainLayout>
    );
  }

  if (error) {
    return (
      <MainLayout>
        <div className="p-8 text-red-600">{error}</div>
      </MainLayout>
    );
  }

  return (
    <MainLayout>
      <div className="max-w-5xl mx-auto px-4 py-8">
        <DashboardTabs activeTab={activeTab} onChange={setActiveTab} showServices={false} />
        <div className="mt-6">
          {activeTab === "requests" && (
            <section>
              <SectionList
                title="Booking Requests"
                data={bookingRequests}
                renderItem={(r) => <BookingRequestCard key={r.id} request={r} onUpdate={() => {}} />}
                emptyState={<span>No requests yet</span>}
              />
            </section>
          )}
          {activeTab === "bookings" && (
            <section>
              <SectionList
                title="Upcoming Bookings"
                data={upcomingBookings}
                emptyState={<span>No bookings yet</span>}
                renderItem={(booking) => (
                  <div key={booking.id} className="bg-white p-4 shadow rounded-lg">
                    <div className="font-medium text-gray-900">
                      {booking.artist?.first_name || "Unknown"} {booking.artist?.last_name || ""}
                    </div>
                    <div className="text-sm text-gray-500">{booking.service?.title || "â€”"}</div>
                    <div className="text-sm text-gray-500">
                      {format(new Date(booking.start_time), "MMM d, yyyy h:mm a")}
                    </div>
                    <div className="mt-2 flex justify-between items-center">
                      <span
                        className={`inline-flex rounded-full px-2 text-xs font-semibold leading-5 ${
                          booking.status === "completed"
                            ? "bg-green-100 text-green-800"
                            : booking.status === "cancelled"
                            ? "bg-red-100 text-red-800"
                            : booking.status === "confirmed"
                            ? "bg-brand-light text-brand-dark"
                            : "bg-yellow-100 text-yellow-800"
                        }`}
                      >
                        {formatStatus(booking.status)}
                      </span>
                      <span className="text-sm text-gray-500">
                        {formatCurrency(Number(booking.total_price))}
                      </span>
                    </div>
                  </div>
                )}
              />
              {bookings.length > upcomingBookings.length && (
                <div className="mt-2">
                  <Link href="/dashboard/client/bookings" className="text-brand-dark hover:underline text-sm">
                    View All Bookings
                  </Link>
                </div>
              )}
            </section>
          )}
        </div>
      </div>
    </MainLayout>
  );
}

