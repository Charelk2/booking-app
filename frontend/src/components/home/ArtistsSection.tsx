'use client';

import {
  useEffect,
  useState,
  useMemo,
  useRef,
} from 'react';
import Link from 'next/link';
import { ChevronRightIcon } from '@heroicons/react/24/solid';
import ServiceProviderCardCompact from '@/components/service-provider/ServiceProviderCardCompact';
import { getServiceProviders } from '@/lib/api';
import { getFullImageUrl } from '@/lib/utils';
import type { ServiceProviderProfile, SearchParams } from '@/types';

interface ArtistsSectionProps {
  title: string;
  query?: Partial<SearchParams>;
  limit?: number;
  hideIfEmpty?: boolean;
}

function CardSkeleton() {
  return (
    <div className="rounded-xl bg-white overflow-hidden animate-pulse">
      <div className="aspect-[4/3] bg-gray-200" />
      <div className="p-3 space-y-1">
        <div className="h-3 bg-gray-200 rounded" />
        <div className="h-2.5 bg-gray-200 rounded w-1/2" />
      </div>
    </div>
  );
}

export default function ArtistsSection({
  title,
  query = {},
  limit = 12,
  hideIfEmpty = false,
}: ArtistsSectionProps) {
  const [artists, setArtists] = useState<ServiceProviderProfile[]>([]);
  const [loading, setLoading] = useState(true);
  const scrollRef = useRef<HTMLDivElement>(null);
  const [canScrollRight, setCanScrollRight] = useState(false);

  const serializedQuery = useMemo(() => JSON.stringify(query), [query]);

  useEffect(() => {
    let isMounted = true;
    async function fetchArtists() {
      setLoading(true);
      try {
        const params = JSON.parse(serializedQuery) as Record<string, unknown>;
        const res = await getServiceProviders({ ...(params as object), limit });
        if (isMounted) {
          setArtists(res.data.filter((a) => a.business_name || a.user));
        }
      } catch (err) {
        console.error(err);
      } finally {
        if (isMounted) {
          setLoading(false);
        }
      }
    }
    fetchArtists();
    return () => {
      isMounted = false;
    };
  }, [serializedQuery, limit]);

  useEffect(() => {
    const updateScrollButtons = () => {
      const el = scrollRef.current;
      if (!el) return;
      setCanScrollRight(el.scrollLeft + el.clientWidth < el.scrollWidth);
    };
    updateScrollButtons();
    const el = scrollRef.current;
    if (!el) return;
    el.addEventListener('scroll', updateScrollButtons);
    window.addEventListener('resize', updateScrollButtons);
    return () => {
      el.removeEventListener('scroll', updateScrollButtons);
      window.removeEventListener('resize', updateScrollButtons);
    };
  }, [artists]);

  const scrollBy = (offset: number) => {
    scrollRef.current?.scrollBy({ left: offset, behavior: 'smooth' });
  };

  if (!loading && artists.length === 0 && hideIfEmpty) {
    return null;
  }

  const seeAllHref = `/search?${new URLSearchParams(query as Record<string, string>).toString()}`;
  const showSeeAll = artists.length === limit;

  return (
    <section className="full-width mx-auto px-4 sm:px-6 lg:px-8 py-10">
      <div className="flex items-end justify-between mb-4">
        <h2 className="text-xl font-semibold text-gray-900">{title}</h2>
        {showSeeAll && (
          <Link href={seeAllHref} className="text-sm text-brand hover:underline">
            See all
          </Link>
        )}
      </div>
      {loading ? (
        <div className="grid grid-cols-2 sm:grid-cols-3 lg:grid-cols-5 xl:grid-cols-7 gap-2 md:gap-2">
          {Array.from({ length: limit }).map((_, i) => (
            <CardSkeleton key={i} />
          ))}
        </div>
      ) : artists.length > 0 ? (
        artists.length > 7 ? (
          <>
            <div className="relative hidden xl:block">
              <div
                ref={scrollRef}
                className="flex items-stretch gap-2 overflow-x-auto scroll-smooth pb-2"
              >
                {artists.map((a) => {
                  const name = a.business_name || `${a.user.first_name} ${a.user.last_name}`;
                  return (
                    <div key={a.id} className="flex-shrink-0 basis-[14.285%]">
                      <ServiceProviderCardCompact
                        serviceProviderId={a.id}
                        name={name}
                        subtitle={a.custom_subtitle || undefined}
                        imageUrl={
                          getFullImageUrl(a.profile_picture_url || a.portfolio_urls?.[0]) || undefined
                        }
                        price={
                          a.hourly_rate && a.price_visible ? Number(a.hourly_rate) : undefined
                        }
                        rating={a.rating ?? undefined}
                        ratingCount={a.rating_count ?? undefined}
                        location={a.location}
                        categories={a.service_categories}
                        href={`/service-providers/${a.id}`}
                        className="h-full w-full"
                      />
                    </div>
                  );
                })}
              </div>
              <button
                type="button"
                aria-label="Next"
                className="absolute right-0 top-1 z-10 rounded-full border bg-white p-2 opacity-50 shadow disabled:opacity-50"
                disabled={!canScrollRight}
                onClick={() => scrollBy(200)}
              >
                <ChevronRightIcon className="h-4 w-4" />
              </button>
            </div>
            <div className="grid grid-cols-2 sm:grid-cols-3 lg:grid-cols-5 xl:hidden gap-2 md:gap-2">
              {artists.map((a) => {
                const name = a.business_name || `${a.user.first_name} ${a.user.last_name}`;
                return (
                  <ServiceProviderCardCompact
                    key={a.id}
                    serviceProviderId={a.id}
                    name={name}
                    subtitle={a.custom_subtitle || undefined}
                    imageUrl={
                      getFullImageUrl(a.profile_picture_url || a.portfolio_urls?.[0]) || undefined
                    }
                    price={
                      a.hourly_rate && a.price_visible ? Number(a.hourly_rate) : undefined
                    }
                    rating={a.rating ?? undefined}
                    ratingCount={a.rating_count ?? undefined}
                    location={a.location}
                    categories={a.service_categories}
                    href={`/service-providers/${a.id}`}
                  />
                );
              })}
            </div>
          </>
        ) : (
          <div className="grid grid-cols-2 sm:grid-cols-3 lg:grid-cols-5 xl:grid-cols-7 gap-2 md:gap-2">
            {artists.map((a) => {
              const name = a.business_name || `${a.user.first_name} ${a.user.last_name}`;
              return (
                <ServiceProviderCardCompact
                  key={a.id}
                  serviceProviderId={a.id}
                  name={name}
                  subtitle={a.custom_subtitle || undefined}
                  imageUrl={
                    getFullImageUrl(a.profile_picture_url || a.portfolio_urls?.[0]) || undefined
                  }
                  price={
                    a.hourly_rate && a.price_visible ? Number(a.hourly_rate) : undefined
                  }
                  rating={a.rating ?? undefined}
                  ratingCount={a.rating_count ?? undefined}
                  location={a.location}
                  categories={a.service_categories}
                  href={`/service-providers/${a.id}`}
                />
              );
            })}
          </div>
        )
      ) : (
        <p>No service providers found.</p>
      )}
    </section>
  );
}
