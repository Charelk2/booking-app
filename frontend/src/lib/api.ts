// frontend/src/lib/api.ts

import axios, { AxiosProgressEvent } from 'axios';
import { extractErrorMessage } from './utils';
import {
  User,
  ArtistProfile,
  Service,
  Booking,
  Review,
  BookingRequestCreate,
  BookingRequest,
  QuoteCreate,
  Quote,
  QuoteV2Create,
  QuoteV2,
  BookingSimple,
  Message,
  MessageCreate,
  SoundProvider,
  ArtistSoundPreference,
  QuoteCalculationResponse,
  Notification,
  ThreadNotification,
} from '@/types';

// Create a single axios instance for all requests
const api = axios.create({
  baseURL: process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000',
  headers: {
    'Content-Type': 'application/json',
  },
});

// Automatically attach the bearer token (if present) to every request
api.interceptors.request.use(
  (config) => {
    if (typeof window !== 'undefined') {
      const token =
        localStorage.getItem('token') || sessionStorage.getItem('token');
      if (token) {
        config.headers.Authorization = `Bearer ${token}`;
      } else if (config.headers && 'Authorization' in config.headers) {
        delete config.headers.Authorization;
      }
    }
    return config;
  },
  (error) => Promise.reject(error)
);

// Provide consistent error messages across the app
api.interceptors.response.use(
  (response) => response,
  (error) => {
    if (axios.isAxiosError(error)) {
      const status = error.response?.status;
      let message = extractErrorMessage(error.response?.data?.detail);

      if (status) {
        const map: Record<number, string> = {
          400: 'Bad request. Please verify your input.',
          401: 'Authentication required. Please log in.',
          403: 'You do not have permission to perform this action.',
          404: 'Resource not found.',
          422: 'Validation failed. Please check your input.',
          500: 'Server error. Please try again later.',
        };
        message = map[status] || message;
      } else {
        message = 'Network error. Please check your connection.';
      }

      console.error('API error:', error);
      return Promise.reject(new Error(message));
    }

    console.error('Unexpected API error:', error);
    return Promise.reject(
      error instanceof Error ? error : new Error('An unexpected error occurred.'),
    );
  },
);

// ─── AUTH (no /api/v1 prefix) ───────────────────────────────────────────────────

export const register = (data: Partial<User>) =>
  api.post('/auth/register', data);

export const login = (email: string, password: string) => {
  const params = new URLSearchParams();
  params.append('username', email);
  params.append('password', password);

  return api.post('/auth/login', params, {
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
  });
};

// ─── All other resources live under /api/v1 ────────────────────────────────────

const API_V1 = '/api/v1';

// Helper to ensure API responses always include `user_id`
const normalizeArtistProfile = (
  profile: Partial<ArtistProfile> | ArtistProfile
): ArtistProfile => ({
  ...profile,
  user_id: (profile.user_id ?? profile.id) as number,
});

// ─── ARTISTS ───────────────────────────────────────────────────────────────────

export const getArtists = async (params?: {
  category?: string;
  location?: string;
  sort?: string;
}) => {
  const res = await api.get<ArtistProfile[]>(`${API_V1}/artist-profiles/`, {
    params,
  });
  return { ...res, data: res.data.map(normalizeArtistProfile) };
};

export const getArtist = async (userId: number) => {
  const res = await api.get<ArtistProfile>(`${API_V1}/artist-profiles/${userId}`);
  return { ...res, data: normalizeArtistProfile(res.data) };
};

export const getArtistAvailability = (artistId: number) =>
  api.get<{ unavailable_dates: string[] }>(`${API_V1}/artist-profiles/${artistId}/availability`);

export const getArtistProfileMe = async () => {
  const res = await api.get<ArtistProfile>(`${API_V1}/artist-profiles/me`);
  return { ...res, data: normalizeArtistProfile(res.data) };
};

export const updateMyArtistProfile = (data: Partial<ArtistProfile>) =>
  api.put(`${API_V1}/artist-profiles/me`, data);

export const uploadMyArtistProfilePicture = (file: File) => {
  const formData = new FormData();
  formData.append('file', file);
  return api.post<ArtistProfile>(
    `${API_V1}/artist-profiles/me/profile-picture`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } }
  );
};

export const uploadMyArtistCoverPhoto = (file: File) => {
  const formData = new FormData();
  formData.append('file', file);
  return api.post<ArtistProfile>(
    `${API_V1}/artist-profiles/me/cover-photo`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } }
  );
};

// ─── SERVICES ──────────────────────────────────────────────────────────────────

// “services by artist” is GET /api/v1/services/artist/{artist_user_id}
export const getArtistServices = (artistUserId: number) =>
  api.get<Service[]>(`${API_V1}/services/artist/${artistUserId}`);

export const getAllServices = () =>
  api.get<Service[]>(`${API_V1}/services/`);

export const getService = (serviceId: number) =>
  api.get<Service>(`${API_V1}/services/${serviceId}`);

// create / update / delete a service: POST /api/v1/services, PUT /api/v1/services/{id}, DELETE /api/v1/services/{id}
export const createService = (data: Partial<Service>) =>
  api.post(`${API_V1}/services/`, data);

export const updateService = (id: number, data: Partial<Service>) =>
  api.put(`${API_V1}/services/${id}`, data);

export const deleteService = (id: number) =>
  api.delete(`${API_V1}/services/${id}`);

// ─── BOOKINGS ──────────────────────────────────────────────────────────────────

// create booking: POST /api/v1/bookings
export const createBooking = (data: Partial<Booking>) =>
  api.post(`${API_V1}/bookings`, data);

// client’s bookings: GET /api/v1/bookings/my-bookings
export const getMyClientBookings = () =>
  api.get<Booking[]>(`${API_V1}/bookings/my-bookings`);

// artist’s bookings: GET /api/v1/bookings/artist-bookings
export const getMyArtistBookings = () =>
  api.get<Booking[]>(`${API_V1}/bookings/artist-bookings`);

// read a single booking: GET /api/v1/bookings/{bookingId}
export const getBookingDetails = (bookingId: number) =>
  api.get<Booking>(`${API_V1}/bookings/${bookingId}`);

// update status: PATCH /api/v1/bookings/{booking_id}/status
export const updateBookingStatus = (id: number, status: Booking['status']) =>
  api.patch(`${API_V1}/bookings/${id}/status`, { status });

// ─── REVIEWS ───────────────────────────────────────────────────────────────────

// create review for a booking: POST /api/v1/reviews/bookings/{booking_id}/reviews
export const createReviewForBooking = (
  bookingId: number,
  data: Partial<Omit<Review, 'booking_id' | 'id'>>
) =>
  api.post(
    `${API_V1}/reviews/bookings/${bookingId}/reviews`,
    data
  );

// list reviews for an artist: GET /api/v1/reviews/artist-profiles/{artist_id}/reviews
export const getArtistReviews = (artistUserId: number) =>
  api.get<Review[]>(
    `${API_V1}/reviews/artist-profiles/${artistUserId}/reviews`
  );

// ─── BOOKING REQUESTS & QUOTES ─────────────────────────────────────────────────

// Create a new booking request (client → artist):
//    POST /api/v1/booking-requests/
// Body must match BookingRequestCreate interface.
export const createBookingRequest = (data: BookingRequestCreate) =>
  api.post<BookingRequest>(`${API_V1}/booking-requests/`, data);

// Optionally, if you want to get a list of booking requests (e.g., for a client dashboard):
export const getMyBookingRequests = () =>
  api.get<BookingRequest[]>(`${API_V1}/booking-requests/me/client`);

// If the artist needs to fetch requests addressed to them:
export const getBookingRequestsForArtist = () =>
  api.get<BookingRequest[]>(`${API_V1}/booking-requests/me/artist`);

// If you want to fetch a single booking request by ID:
export const getBookingRequestById = (id: number) =>
  api.get<BookingRequest>(`${API_V1}/booking-requests/${id}`);

// Update an existing booking request as the client
export const updateBookingRequest = (
  id: number,
  data: Partial<BookingRequestCreate> & { status?: string }
) => api.put<BookingRequest>(`${API_V1}/booking-requests/${id}/client`, data);

// Create a new quote (artist → client) for an existing booking request:
//    POST /api/v1/quotes/
// Body must match QuoteCreate interface.
export const createQuoteForRequest = (
  requestId: number,
  data: QuoteCreate
) => api.post<Quote>(`${API_V1}/booking-requests/${requestId}/quotes`, data);

// Optionally, fetch all quotes for a given booking request:
export const getQuotesForBookingRequest = (bookingRequestId: number) =>
  api.get<Quote[]>(
    `${API_V1}/booking-requests/${bookingRequestId}/quotes`
  );

export const createQuoteV2 = (data: QuoteV2Create) =>
  api.post<QuoteV2>(`${API_V1}/quotes`, data);

export const getQuoteV2 = (quoteId: number) =>
  api.get<QuoteV2>(`${API_V1}/quotes/${quoteId}`);

export const acceptQuoteV2 = (quoteId: number) =>
  api.post<BookingSimple>(`${API_V1}/quotes/${quoteId}/accept`, {});

// ─── MESSAGES ───────────────────────────────────────────────────────────
export const getMessagesForBookingRequest = (bookingRequestId: number) =>
  api.get<Message[]>(
    `${API_V1}/booking-requests/${bookingRequestId}/messages`
  );

export const postMessageToBookingRequest = (
  bookingRequestId: number,
  data: MessageCreate
) =>
  api.post<Message>(
    `${API_V1}/booking-requests/${bookingRequestId}/messages`,
    data
  );

export const uploadMessageAttachment = (
  bookingRequestId: number,
  file: File,
  onUploadProgress?: (event: AxiosProgressEvent) => void,
) => {
  const formData = new FormData();
  formData.append('file', file);
  return api.post<{ url: string }>(
    `${API_V1}/booking-requests/${bookingRequestId}/attachments`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' }, onUploadProgress }
  );
};

export const uploadBookingAttachment = (
  formData: FormData,
  onUploadProgress?: (event: AxiosProgressEvent) => void,
) =>
  api.post<{ url: string }>(
    `${API_V1}/booking-requests/attachments`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' }, onUploadProgress }
  );


// ─── SOUND PROVIDERS ─────────────────────────────────────────────────────────
export const getSoundProviders = () =>
  api.get<SoundProvider[]>(`${API_V1}/sound-providers/`);

export const createSoundProvider = (data: Partial<SoundProvider>) =>
  api.post<SoundProvider>(`${API_V1}/sound-providers/`, data);

export const updateSoundProvider = (
  id: number,
  data: Partial<SoundProvider>
) => api.put<SoundProvider>(`${API_V1}/sound-providers/${id}`, data);

export const deleteSoundProvider = (id: number) =>
  api.delete(`${API_V1}/sound-providers/${id}`);

export const getSoundProvidersForArtist = (artistId: number) =>
  api.get<ArtistSoundPreference[]>(`${API_V1}/sound-providers/artist/${artistId}`);

export const addArtistSoundPreference = (
  artistId: number,
  data: { provider_id: number; priority?: number }
) =>
  api.post<ArtistSoundPreference>(`${API_V1}/sound-providers/artist/${artistId}`,
    data);

// ─── QUOTE CALCULATOR ───────────────────────────────────────────────────────
export const calculateQuote = (params: {
  base_fee: number;
  distance_km: number;
  provider_id?: number;
  accommodation_cost?: number;
}) => api.post<QuoteCalculationResponse>(`${API_V1}/quotes/calculate`, params);

// ─── NOTIFICATIONS ───────────────────────────────────────────────────────────
export const getNotifications = (skip = 0, limit = 20) =>
  api.get<Notification[]>(
    `${API_V1}/notifications?skip=${skip}&limit=${limit}`,
  );

export const getGroupedNotifications = () =>
  api.get<Record<string, Notification[]>>(`${API_V1}/notifications/grouped`);

export const markNotificationRead = (id: number) =>
  api.put<Notification>(`${API_V1}/notifications/${id}/read`);

export const markAllNotificationsRead = () =>
  api.put<{ updated: number }>(`${API_V1}/notifications/read-all`);

export const getMessageThreads = () =>
  api.get<ThreadNotification[]>(`${API_V1}/notifications/message-threads`);

export const markThreadRead = (bookingRequestId: number) =>
  api.put(`${API_V1}/notifications/message-threads/${bookingRequestId}/read`);

export default api;
