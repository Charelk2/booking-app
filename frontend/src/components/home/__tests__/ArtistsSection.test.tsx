import { flushPromises } from "@/test/utils/flush";
import { createRoot } from 'react-dom/client';
import React from 'react';
import { act } from 'react-dom/test-utils';
import ArtistsSection from '../ArtistsSection';
import * as api from '@/lib/api';
import type { ArtistProfile } from '@/types';

jest.mock('@/lib/api');

function setup() {
  const container = document.createElement('div');
  document.body.appendChild(container);
  const root = createRoot(container);
  return { container, root };
}

