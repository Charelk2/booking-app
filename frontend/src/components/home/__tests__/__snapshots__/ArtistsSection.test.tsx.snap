// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistsSection matches snapshot 1`] = `
<section
  class="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8 py-12"
>
  <div
    class="flex items-end justify-between mb-6"
  >
    <h2
      class="text-2xl font-semibold text-gray-900"
    >
      Demo
    </h2>
  </div>
  <div
    class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-6"
  >
    <div
      class="rounded-2xl bg-white shadow-lg overflow-hidden transition-shadow"
    >
      <a
        href="/artists/1"
      >
        <div
          class="relative h-48 w-full overflow-hidden rounded-t-2xl bg-gray-100"
        >
          <div
            class="absolute inset-0 animate-pulse bg-gray-200"
          />
          <img
            alt="A"
            class="object-cover w-full h-full"
            data-nimg="1"
            decoding="async"
            height="512"
            loading="lazy"
            src="/default-avatar.svg"
            style="color: transparent;"
            width="512"
          />
          <div
            class="absolute inset-0 flex items-center justify-center bg-black/60 opacity-0 group-hover:opacity-100 transition"
          >
            <button
              class="text-sm bg-brand text-white px-4 py-1.5 rounded-md focus:outline-none focus-visible:ring"
              type="button"
            >
              Book Now
            </button>
          </div>
        </div>
      </a>
      <div
        class="flex flex-col flex-1 p-6"
      >
        <div
          class="flex items-center"
        >
          <h2
            class="flex-1 text-lg font-semibold text-gray-900 truncate mt-4 mb-2"
          >
            A
          </h2>
        </div>
        <hr
          class="mt-4 mb-4 border-gray-200"
        />
        <div
          class="flex justify-between items-center text-sm text-gray-700"
        >
          <span
            class="flex items-center"
          >
            <svg
              aria-hidden="true"
              class="h-4 w-4 mr-1 text-yellow-400"
              data-slot="icon"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M10.788 3.21c.448-1.077 1.976-1.077 2.424 0l2.082 5.006 5.404.434c1.164.093 1.636 1.545.749 2.305l-4.117 3.527 1.257 5.273c.271 1.136-.964 2.033-1.96 1.425L12 18.354 7.373 21.18c-.996.608-2.231-.29-1.96-1.425l1.257-5.273-4.117-3.527c-.887-.76-.415-2.212.749-2.305l5.404-.434 2.082-5.005Z"
                fill-rule="evenodd"
              />
            </svg>
            <span
              class="text-gray-400"
            >
              No ratings yet
            </span>
          </span>
        </div>
        <div
          class="flex justify-between items-center mt-4"
        >
          <span />
          <a
            href="/artists/1"
          >
            <button
              class="bg-brand text-white text-sm px-4 py-1.5 rounded-md hover:bg-brand-dark"
              type="button"
            >
              View Profile
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
</section>
`;
